roscisz/TensorHive
# Mailbot - new violation handler

Read all sections carefully :tea: 

## Features
* Mailbot sends email every `interval` minutes (each email address is timed separately)
* Admin can also get emails about violations (useful, when intruder does not have TH account or uses invalid/fake email address)
* Admin and intruder can get different email bodies
* There is a body templating mechanism that supports HTML and is fully customizable via `mailbot_config.ini`

## Preview
### Email to intruder
![image](https://user-images.githubusercontent.com/12485656/51789201-af2ff200-2186-11e9-946f-16d993835868.png)

### Email to admin:
![image](https://user-images.githubusercontent.com/12485656/51789016-b3f3a680-2184-11e9-9708-b99d3102931f.png)

### Email HTML templating:
![image](https://user-images.githubusercontent.com/12485656/51789022-bce47800-2184-11e9-9abf-d85e7cc15b6e.png)


## Configuration
* Uses separate config file - `mail_config.ini`
![image](https://user-images.githubusercontent.com/12485656/51788898-28c5e100-2183-11e9-9448-ced590f59366.png)
* `mail_config.ini` will be always recreated (identically as other config files)
* Mailbot's SMTP credentials must be defined by 2 environment variables (but the names can be customized in config file)
* User will know if these env vars are undeclared
![image](https://user-images.githubusercontent.com/12485656/51788895-1ba8f200-2183-11e9-9b65-fb7df5e14dd4.png)

## Manual testing scenarios
- Make clean install of `tensorhive` (remove everything under `~/.config/Tensorhive/`)
- Remember to fill in `hosts_config.ini`
- Notice warning about undeclared env variables,
```bash
export TENSORHIVE_MAILBOT_EMAIL="ask @micmarty"
export TENSORHIVE_MAILBOT_PASSWORD="ask @micmarty"
```
- Create accounts:
`tensorhive create user`
- create admin account
- create intruder account
- Make reservation as admin
- Run GPU process as intruder
- Check intruder's inbox (and optionally admin's - `notify_admin = yes`)

### More:
- experiment with `mailbot_config.ini` to find bugs
- look for edge cases

## Notes
@roscisz 
Code used to be clean at some point, now it looks clunky, especially for me/you.
I'll do a hard refactor, because that code will be hard to maintain/extend later.
I have a concept in my mind already, so it's a matter of time and priorities.
What's your opinion?
